README.md
# Cardano Node Ui

<p align="left">
<img alt="Test" src="https://img.shields.io/github/actions/workflow/status/fabianbormann/cardano-node-ui/test.yml?label=Test&style=for-the-badge" />
<img alt="Release" src="https://img.shields.io/github/actions/workflow/status/fabianbormann/cardano-node-ui/release.yml?label=Release&style=for-the-badge" />
<img alt="Build" src="https://img.shields.io/github/actions/workflow/status/fabianbormann/cardano-node-ui/build.yml?label=Build&style=for-the-badge" />
<a href="https://conventionalcommits.org"><img alt="conventionalcommits" src="https://img.shields.io/badge/Conventional%20Commits-1.0.0-%23FE5196?logo=conventionalcommits&style=for-the-badge" /></a>
</p>

This project aims to provide a streamlined, single-click ui solution for launching a Cardano node

<img src="https://user-images.githubusercontent.com/1525818/232276588-89c5288d-42a6-4820-975b-7e2c25cf75da.png" width="75%"/>

## 📌 Download Latest Release

📦 Cardano Node Ui v0.1.6

<div><a href="https://github.com/fabianbormann/cardano-node-ui/releases/download/v0.1.6/cardano-node-ui-0.1.6.dmg">⬇️ Download | OS X 🍎</a></div>
<div><a href="https://github.com/fabianbormann/cardano-node-ui/releases/download/v0.1.6/cardano-node-ui-0.1.6.AppImage">⬇️ Download | Linux 🐧</a></div>
<div><a href="https://github.com/fabianbormann/cardano-node-ui/releases/download/v0.1.6/cardano-node-ui-Setup-0.1.6.exe">⬇️ Download | Windows 🪟🪟</a></div>

## ❤️‍🔥 Motivation

Currently, there is no graphical user interface to start and monitor a Cardano node.

For some testing or development purposes, a personal Cardano node is essential. However, without a simple and streamlined user interface, users must rely on command line tools to set up and manage nodes, which can be time consuming. This can be time consuming. It requires reading lengthy documentation and executing a series of commands in a terminal.

This project aims to provide a one-click solution for starting a Cardano node and monitoring its synchronization progress. By offering an intuitive and easy-to-use interface, we hope to make it easier for individuals and organizations to participate in the Cardano network and contribute to its growth.

## 💪 Contributing

Thanks for considering contributing to cardano-node-ui! Your help is greatly appreciated!

If you would like to contribute, please follow these steps:

- Fork this repository.
- Create a new branch with your changes.
- Submit a pull request for review.
- We appreciate all contributions and thank you for your support.

🙏 For more information, please checkout our [contributing guidelines](./CONTRIBUTING.md).


CHANGELOG.md
# Changelog

## [0.1.6](https://github.com/fabianbormann/cardano-node-ui/compare/v0.1.5...v0.1.6) (2023-05-18)


### Features

* disable button while restarting ([4d53d22](https://github.com/fabianbormann/cardano-node-ui/commit/4d53d22428a3e98b6b00384bee8ab5fcb790624d))

## [0.1.5](https://github.com/fabianbormann/cardano-node-ui/compare/v0.1.4...v0.1.5) (2023-05-07)


### Features

* add dashboard test ([85c4ca4](https://github.com/fabianbormann/cardano-node-ui/commit/85c4ca4f3cd2cf46b55710e604ad9d290938e821))

## [0.1.4](https://github.com/fabianbormann/cardano-node-ui/compare/v0.1.3...v0.1.4) (2023-04-16)


### Bug Fixes

* define a relative project root to ensure that there is no empty window when opening the app ([8f6a58a](https://github.com/fabianbormann/cardano-node-ui/commit/8f6a58a3ba8c21edcdf37d7db1cb225a778d0b8d))

## [0.1.3](https://github.com/fabianbormann/cardano-node-ui/compare/v0.1.2...v0.1.3) (2023-04-15)


### Features

* add copy field for node.socket path ([f36dbdb](https://github.com/fabianbormann/cardano-node-ui/commit/f36dbdb0242491c8b271dbe0515155b85ff695f7))


### Bug Fixes

* prevent increasing background speed ([261c609](https://github.com/fabianbormann/cardano-node-ui/commit/261c609a8513ecab0db631e85d7e8611f93ac26c))
* the app will close if clicking the close button ([d6d66ac](https://github.com/fabianbormann/cardano-node-ui/commit/d6d66ac9a37d23fef4eb2e1488c0d473f2ce8d96))

## [0.1.2](https://github.com/fabianbormann/cardano-node-ui/compare/v0.1.1...v0.1.2) (2023-04-12)


### Features

* add ui button for network selection and update node binary ([458291f](https://github.com/fabianbormann/cardano-node-ui/commit/458291fc05827f2ce243daab69c6f46276aae91b))


### Bug Fixes

* repair build action ([47592c0](https://github.com/fabianbormann/cardano-node-ui/commit/47592c08857b775db23f07a5a991ed56a1426245))


### Miscellaneous Chores

* release 0.1.1 ([0530d47](https://github.com/fabianbormann/cardano-node-ui/commit/0530d47fb0187242a6faaa3efba49b2b5678c604))
* release 0.1.2 ([09c12d0](https://github.com/fabianbormann/cardano-node-ui/commit/09c12d0108164146d6f8b849ca65793ff6fa9343))

## [0.1.1](https://github.com/fabianbormann/cardano-node-ui/compare/v0.1.0...v0.1.1) (2023-04-10)


### Bug Fixes

* remove warnings and check npm build ([57a351a](https://github.com/fabianbormann/cardano-node-ui/commit/57a351aa38b7cdce0b27d2f3ddddd1228f685459))

## 0.1.0 (2023-04-10)


### Features

* add basic ui and animated background ([80f5102](https://github.com/fabianbormann/cardano-node-ui/commit/80f51027aa5f9bd7bbbdc0e205e43401fcd1600f))
* display download progress ([349998f](https://github.com/fabianbormann/cardano-node-ui/commit/349998fd819cff8c4b3ea3ce7b19b43bd6a84cd5))
* display node sync progress ([27d1936](https://github.com/fabianbormann/cardano-node-ui/commit/27d1936a9aef59cdea0effbf1b4353ced93b5158))


### Bug Fixes

* configure release please to produce 0.x.x versions ([691ef1a](https://github.com/fabianbormann/cardano-node-ui/commit/691ef1a0dba1ac363c84ac5487e66347142ab246))
* remove test entry from github action ([3270d06](https://github.com/fabianbormann/cardano-node-ui/commit/3270d061098270715ee09d10676adaf53ef4677c))


public/preload.js
const { contextBridge, ipcRenderer } = require('electron');

contextBridge.exposeInMainWorld('electron', {
  closeApp: () => {
    return ipcRenderer.sendSync('close-app');
  },
  getSocketPath: (directory, network) => {
    return ipcRenderer.sendSync('get-socket-path', directory, network);
  },
  openUrlInBrowser: (url) => {
    ipcRenderer.send('open-external-url', url);
  },
  openDialog: () => {
    return ipcRenderer.sendSync('open-dialog');
  },
  addEventListener: (channel, callback) => {
    ipcRenderer.on(channel, (_, data) => callback(data));
  },
  removeAllListeners: (channel) => {
    ipcRenderer.removeAllListeners(channel);
  },
  getDefaultDirectory: () => ipcRenderer.invoke('getDefaultPath'),
  startNode: (directory, network) =>
    ipcRenderer.send('start-node', directory, network),
  stopNode: () => ipcRenderer.send('stop-node'),
});


src/components/Particle.ts
export default class Particle {
  private canvas: HTMLCanvasElement;
  private context: CanvasRenderingContext2D;
  private color: string;
  x: number;
  y: number;
  velocity: { x: number; y: number };

  constructor(
    canvas: HTMLCanvasElement,
    context: CanvasRenderingContext2D,
    options: { color: string; velocity: number }
  ) {
    this.canvas = canvas;
    this.context = context;
    this.color = options.color;

    this.x = Math.random() * this.canvas.width;
    this.y = Math.random() * this.canvas.height;
    this.velocity = {
      x: (Math.random() - 0.5) * options.velocity,
      y: (Math.random() - 0.5) * options.velocity,
    };
  }

  update() {
    if (this.x > this.canvas.width + 20 || this.x < -20) {
      this.velocity.x = -this.velocity.x;
    }
    if (this.y > this.canvas.height + 20 || this.y < -20) {
      this.velocity.y = -this.velocity.y;
    }
    this.x += this.velocity.x;
    this.y += this.velocity.y;
  }

  draw() {
    this.context.beginPath();
    this.context.fillStyle = this.color;
    this.context.globalAlpha = 0.7;
    this.context.arc(this.x, this.y, 1.5, 0, 2 * Math.PI);
    this.context.fill();
  }
}


cypress/e2e/dashboard.spec.cy.ts
/// <reference types="cypress" />

describe('dashboard spec', () => {
  before(() => {
    cy.visit('http://localhost:3000').then(() =>
      cy.window().then((window) => {
        (window as any).electron = {
          getDefaultDirectory: async () => {
            return '/Users/suomynona/.cardano-node-ui';
          },
          getSocketPath: (directory: string, selectedNetwork: string) => {
            const databaseDirectory = `${directory}/${selectedNetwork}-db`;
            const socketPath = `${databaseDirectory}/node.socket`;

            return `export CARDANO_NODE_SOCKET_PATH="${socketPath}"`;
          },
        };
      })
    );
  });

  it('should switch the network and adjust the node.socket path', () => {
    cy.get('[data-testid="network-selector-input"]').should(
      'have.value',
      'mainnet'
    );
    cy.get('[data-testid="socket-path-field"]')
      .find('input')
      .invoke('val')
      .should('include', 'mainnet-db');

    cy.get('[data-testid="network-selector"]').click();
    cy.get('[data-testid="network-option-preprod"]').click({ force: true });

    cy.get('[data-testid="network-selector-input"]').should(
      'have.value',
      'preprod'
    );
    cy.get('[data-testid="socket-path-field"]')
      .find('input')
      .invoke('val')
      .should('include', 'preprod-db');
  });
});


src/components/ParticleNetwork.ts
import {
  ContainerSize,
  densityOptions,
  ParticleNetworkOptions,
} from '../global/types';
import Particle from './Particle';

export default class ParticleNetwork {
  container: HTMLDivElement;
  private containerSize: ContainerSize;
  private options: ParticleNetworkOptions;
  private canvas: HTMLCanvasElement;
  private context: CanvasRenderingContext2D | null;
  private resetTimer?: number;
  private particles: Array<Particle> = [];
  private mouseParticle: Particle | null = null;
  private removeResizeListener: Function = () => {};
  private removeMouseMoveListener: Function = () => {};
  private removeMouseUpListener: Function = () => {};

  constructor(
    container: HTMLDivElement,
    canvas: HTMLCanvasElement,
    options?: ParticleNetworkOptions
  ) {
    this.container = container;
    this.containerSize = {
      width: this.container.offsetWidth,
      height: this.container.offsetHeight,
    };

    this.options = {
      color: options?.color || '#fff',
      background: options?.background || '#1a252f',
      interactive:
        typeof options?.interactive === 'boolean' ? options.interactive : true,
      velocity: options?.velocity || 0.5,
      useWindowForMouseEvents:
        typeof options?.useWindowForMouseEvents === 'boolean'
          ? options.useWindowForMouseEvents
          : false,
    };
    this.setDensity(options?.density);

    this.canvas = canvas;
    this.context = this.canvas.getContext('2d');
    this.canvas.width = this.containerSize.width;
    this.canvas.height = this.containerSize.height - 5;
    this.init();
  }

  clear() {
    this.options.velocity = 0;
    this.removeMouseMoveListener();
    this.removeMouseUpListener();
    this.removeResizeListener();
    this.particles = [];
    this.mouseParticle = null;
  }

  init() {
    this.setStyles(this.container, {
      background: this.options.background!,
    });

    const resizeListener = () => {
      if (
        this.container.offsetWidth === this.containerSize.width &&
        this.container.offsetHeight === this.containerSize.height
      ) {
        return false;
      }

      this.canvas.width = this.containerSize.width = this.container.offsetWidth;
      this.canvas.height =
        (this.containerSize.height = this.container.offsetHeight) - 5;

      clearTimeout(this.resetTimer);
      this.resetTimer = window.setTimeout(() => {
        this.particles = [];
        for (
          var i = 0;
          i < (this.canvas.width * this.canvas.height) / this.options.density!;
          i++
        ) {
          this.particles.push(
            new Particle(this.canvas, this.context!, {
              color: this.options.color!,
              velocity: this.options.velocity!,
            })
          );
        }
        if (this.options.interactive) {
          this.particles.push(this.mouseParticle!);
        }
      }, 500);
    };

    window.addEventListener('resize', resizeListener);
    this.removeResizeListener = () => {
      window.removeEventListener('resize', resizeListener);
    };

    this.particles = [];
    for (
      var i = 0;
      i < (this.canvas.width * this.canvas.height) / this.options.density!;
      i++
    ) {
      this.particles.push(
        new Particle(this.canvas, this.context!, {
          color: this.options.color!,
          velocity: this.options.velocity!,
        })
      );
    }

    if (this.options.interactive) {
      this.mouseParticle = new Particle(this.canvas, this.context!, {
        color: this.options.color!,
        velocity: this.options.velocity!,
      });
      this.mouseParticle.velocity = {
        x: 0,
        y: 0,
      };
      this.particles.push(this.mouseParticle);

      const mousemoveListener = (event: MouseEvent) => {
        this.mouseParticle!.x = event.clientX - this.canvas.offsetLeft;
        this.mouseParticle!.y = event.clientY - this.canvas.offsetTop;
      };

      if (this.options.useWindowForMouseEvents) {
        window.addEventListener('mousemove', mousemoveListener);
        this.removeMouseMoveListener = () => {
          window.removeEventListener('mousemove', mousemoveListener);
        };
      } else {
        this.canvas.addEventListener('mousemove', mousemoveListener);
        this.removeMouseMoveListener = () => {
          this.canvas.removeEventListener('mousemove', mousemoveListener);
        };
      }

      const mouseupListener = (event: MouseEvent) => {
        this.mouseParticle!.velocity = {
          x: (Math.random() - 0.5) * this.options.velocity!,
          y: (Math.random() - 0.5) * this.options.velocity!,
        };
        this.mouseParticle = new Particle(this.canvas, this.context!, {
          color: this.options.color!,
          velocity: this.options.velocity!,
        });
        this.mouseParticle.velocity = {
          x: 0,
          y: 0,
        };
        this.particles.push(this.mouseParticle);
      };

      if (this.options.useWindowForMouseEvents) {
        window.addEventListener('mouseup', mouseupListener);
        this.removeMouseUpListener = () => {
          window.removeEventListener('mousemove', mouseupListener);
        };
      } else {
        this.canvas.addEventListener('mouseup', mouseupListener);
        this.removeMouseUpListener = () => {
          this.canvas.removeEventListener('mousemove', mouseupListener);
        };
      }
    }

    requestAnimationFrame(this.update.bind(this));
  }

  update() {
    this.context!.clearRect(0, 0, this.canvas.width, this.canvas.height);
    this.context!.globalAlpha = 1;

    for (var i = 0; i < this.particles.length; i++) {
      this.particles[i].update();
      this.particles[i].draw();

      for (var j = this.particles.length - 1; j > i; j--) {
        var distance = Math.sqrt(
          Math.pow(this.particles[i].x - this.particles[j].x, 2) +
            Math.pow(this.particles[i].y - this.particles[j].y, 2)
        );
        if (distance > 120) {
          continue;
        }

        this.context!.beginPath();
        this.context!.strokeStyle = this.options.color!;
        this.context!.globalAlpha = (120 - distance) / 120;
        this.context!.lineWidth = 0.7;
        this.context!.moveTo(this.particles[i].x, this.particles[i].y);
        this.context!.lineTo(this.particles[j].x, this.particles[j].y);
        this.context!.stroke();
      }
    }

    if (this.options.velocity !== 0) {
      requestAnimationFrame(this.update.bind(this));
    }
  }

  setDensity(density?: densityOptions | number) {
    if (typeof density === 'undefined') {
      this.options.density = 10000;
    } else if (typeof density === 'number') {
      this.options.density = density;
    } else if (density === 'high') {
      this.options.density = 5000;
    } else if (density === 'low') {
      this.options.density = 20000;
    } else if (density === 'medium') {
      this.options.density = 10000;
    } else {
      throw Error(
        `This density parameter is not allowed: ${density}. Use a number or the strings 'high' || 'medium' || 'low' instead`
      );
    }
  }

  setStyles(div: HTMLElement, styles: { [key: string]: string }) {
    for (const style in styles) {
      div.style.setProperty(style, styles[style]);
    }
  }
}


public/electron.js
const {
  app,
  BrowserWindow,
  nativeImage,
  ipcMain,
  dialog,
  shell,
} = require('electron');
const path = require('path');
const isDev = require('electron-is-dev');
const Downloader = require('nodejs-file-downloader');
const spawn = require('child_process').spawn;
const fs = require('fs/promises');
const ip = require('ip');

let nodeInstance = null;
let intervalId = -1;
let nodeRunning = false;

class NodeAlreadyRunningError extends Error {}

const networks = {
  mainnet: {
    protocolMagic: 764824073,
  },
  preprod: {
    protocolMagic: 1,
  },
  preview: {
    protocolMagic: 2,
  },
};

const nodeBinaries = {
  darwin: 'cardano-node-1.35.7-macos.tar.gz',
  win32: 'cardano-node-1.35.7-win64.zip',
  linux: 'cardano-node-1.35.7-linux.tar.gz',
};

const nodeBaseUrl =
  'https://update-cardano-mainnet.iohk.io/cardano-node-releases/';

const downloadNodeBinary = async (directory, feedbackFunction) => {
  const nodeBinary = nodeBinaries[process.platform];
  if (nodeBinary) {
    const downloader = new Downloader({
      url: `${nodeBaseUrl}${nodeBinary}`,
      directory: directory,
      cloneFiles: false,
      skipExistingFileName: true,
      onProgress: (percentage) => {
        feedbackFunction(
          `download cardano-node binary ${Math.round(percentage)}%`
        );
      },
    });

    try {
      await downloader.download();
    } catch (error) {
      console.log('Download failed', error);
      throw error;
    }
  }
};

const unpackArchive = (directory) =>
  new Promise((resolve, reject) => {
    const nodeBinary = nodeBinaries[process.platform];
    const binaryPath = path.join(
      directory,
      nodeBinary.replace('.tar.gz', '').replace('.zip', '')
    );

    fs.mkdir(binaryPath, { recursive: true })
      .then(() => {
        const child = spawn('tar', [
          '-xzf',
          path.join(directory, nodeBinary),
          '-C',
          binaryPath,
        ]);
        child.on('close', (code) => {
          console.log(`child process exited with code ${code}`);
          if (code === 0) {
            resolve();
          } else {
            reject();
          }
        });

        child.stdout.on('data', (data) => {
          console.log(`stdout: ${data}`);
        });

        child.stderr.on('data', (data) => {
          console.error(`stderr: ${data}`);
        });
      })
      .catch((error) => reject(error));
  });

const startNode = (directory, network) =>
  new Promise((resolve, reject) => {
    if (nodeInstance !== null) {
      throw new NodeAlreadyRunningError();
    }

    const nodeBinary = nodeBinaries[process.platform];
    const configDirectory = path.join(directory, `${network}-config`);
    const databaseDirectory = path.join(directory, `${network}-db`);
    const binaryPath = path.join(
      directory,
      nodeBinary.replace('.tar.gz', '').replace('.zip', '')
    );

    nodeInstance = spawn(path.join(binaryPath, 'cardano-node'), [
      'run',
      '--topology',
      path.join(configDirectory, 'topology.json'),
      '--database-path',
      databaseDirectory,
      '--socket-path',
      path.join(databaseDirectory, 'node.socket'),
      '--host-addr',
      ip.address(),
      '--port',
      '3001',
      '--config',
      path.join(configDirectory, 'config.json'),
    ]);

    nodeInstance.on('error', (error) => {
      console.log(`child process exited with error ${error.name}`);
      console.log(error.message);
      reject();
    });

    nodeInstance.on('spawn', () => resolve());

    nodeInstance.stdout.on('data', (data) => {
      console.log(`stdout: ${data}`);
    });

    nodeInstance.stderr.on('data', (data) => {
      console.error(`stderr: ${data}`);
    });
  });

const downloadConfig = async (directory, network) => {
  const configDirectory = path.join(directory, `${network}-config`);
  const databaseDirectory = path.join(directory, `${network}-db`);
  await fs.mkdir(configDirectory, { recursive: true });
  await fs.mkdir(databaseDirectory, { recursive: true });

  const filenames = [
    `https://book.world.dev.cardano.org/environments/${network}/config.json`,
    `https://book.world.dev.cardano.org/environments/${network}/db-sync-config.json`,
    `https://book.world.dev.cardano.org/environments/${network}/submit-api-config.json`,
    `https://book.world.dev.cardano.org/environments/${network}/topology.json`,
    `https://book.world.dev.cardano.org/environments/${network}/byron-genesis.json`,
    `https://book.world.dev.cardano.org/environments/${network}/shelley-genesis.json`,
    `https://book.world.dev.cardano.org/environments/${network}/alonzo-genesis.json`,
  ];

  for (const filename of filenames) {
    const downloader = new Downloader({
      url: filename,
      cloneFiles: false,
      directory: configDirectory,
    });
    try {
      const { filePath, downloadStatus } = await downloader.download();
      console.log('All done');
    } catch (error) {
      console.log('Download failed', error);
      throw error;
    }
  }
};

function createWindow() {
  const image = nativeImage.createFromPath(
    path.join(__dirname, 'icons', 'icon.png')
  );

  const mainWindow = new BrowserWindow({
    width: 800,
    height: 600,
    minWidth: 760,
    minHeight: 500,
    title: 'Cardano Node UI',
    frame: false,
    icon: image,
    webPreferences: {
      preload: path.join(__dirname, 'preload.js'),
    },
  });

  mainWindow.setMenu(null);
  mainWindow.loadURL(
    isDev
      ? 'http://localhost:3000'
      : `file://${path.join(__dirname, 'index.html')}`
  );
  mainWindow.maximize();
  return mainWindow;
}

app.whenReady().then(() => {
  const mainWindow = createWindow();

  app.on('activate', function () {
    if (BrowserWindow.getAllWindows().length === 0) createWindow();
  });

  ipcMain.on('open-external-url', (_event, url) => {
    shell.openExternal(url);
  });

  ipcMain.on('close-app', () => {
    app.quit();
  });

  ipcMain.on('open-dialog', (event) => {
    event.returnValue = dialog.showOpenDialogSync({
      properties: ['openDirectory', 'createDirectory', 'promptToCreate'],
    });
  });

  ipcMain.on('get-socket-path', (event, directory, network) => {
    const databaseDirectory = path.join(directory, `${network}-db`);
    const socketPath = path.join(databaseDirectory, 'node.socket');

    if (process.platform === 'win32') {
      event.returnValue = `setx CARDANO_NODE_SOCKET_PATH "${socketPath}"`;
    } else {
      event.returnValue = `export CARDANO_NODE_SOCKET_PATH="${socketPath}"`;
    }
  });

  ipcMain.on('stop-node', () => {
    if (nodeRunning) {
      nodeRunning = false;
      mainWindow.webContents.send('node-status', {
        id: -1,
        timestamp: new Date().getTime(),
        status: 'shutdown',
        message: 'stopping the node ⏸ #{...}',
      });
    }
  });

  ipcMain.on('start-node', async (_event, directory, network = 'mainnet') => {
    nodeRunning = true;

    mainWindow.webContents.send('node-status', {
      id: 0,
      timestamp: new Date().getTime(),
      status: 'download',
      message: 'download cardano-node binary #{...}',
    });

    try {
      const updateDownloadStatus = (message) => {
        mainWindow.webContents.send('node-status', {
          id: 0,
          timestamp: new Date().getTime(),
          status: 'download',
          message: message,
        });
      };

      await downloadNodeBinary(directory, updateDownloadStatus);
    } catch (error) {
      console.error(error);
      mainWindow.webContents.send('node-status', {
        id: -1,
        timestamp: new Date().getTime(),
        status: 'error',
        message: `error while downloading the cardano-node binary #{:x:}`,
      });
      return;
    }

    if (!nodeRunning) {
      mainWindow.webContents.send('node-status', {
        id: -1,
        timestamp: new Date().getTime(),
        status: 'idle',
        message: 'the node has been successfully stopped ✅!',
      });

      return;
    }

    mainWindow.webContents.send('node-status', {
      id: 0,
      timestamp: new Date().getTime(),
      status: 'running',
      message:
        'cardano-node binary binary download completed #{:white_check_mark:}',
    });
    mainWindow.webContents.send('node-status', {
      id: 1,
      timestamp: new Date().getTime(),
      status: 'download',
      message: 'download configuration and node topology #{...}',
    });
    await downloadConfig(directory, network);
    if (!nodeRunning) {
      mainWindow.webContents.send('node-status', {
        id: -1,
        timestamp: new Date().getTime(),
        status: 'idle',
        message: 'the node has been successfully stopped ✅!',
      });
      return;
    }
    mainWindow.webContents.send('node-status', {
      id: 1,
      timestamp: new Date().getTime(),
      status: 'running',
      message:
        'configuration and node topology download completed #{:white_check_mark:}',
    });

    try {
      await unpackArchive(directory);
    } catch (error) {
      console.error(error);
      mainWindow.webContents.send('node-status', {
        id: -1,
        timestamp: new Date().getTime(),
        status: 'error',
        message: `error while unpacking the cardano-node binary #{:x:}`,
      });
      return;
    }

    mainWindow.webContents.send('node-status', {
      id: -1,
      timestamp: new Date().getTime(),
      status: 'running',
      message: `cardano-node binary has been successfully unpacked #{:white_check_mark:}`,
    });

    mainWindow.webContents.send('node-status', {
      id: 2,
      timestamp: new Date().getTime(),
      status: 'running',
      message: 'node initialization running  #{...}',
    });

    try {
      await startNode(directory, network);
    } catch (error) {
      if (error instanceof NodeAlreadyRunningError) {
        mainWindow.webContents.send('node-status', {
          id: 2,
          timestamp: new Date().getTime(),
          status: 'running',
          message: 'node already running',
        });
      } else {
        mainWindow.webContents.send('node-status', {
          id: 2,
          timestamp: new Date().getTime(),
          status: 'error',
          message: `failed to start node${
            typeof error.message !== 'undefined' && ': ' + error.message
          }`,
        });
      }
      return;
    }

    if (nodeRunning) {
      mainWindow.webContents.send('node-status', {
        id: 2,
        timestamp: new Date().getTime(),
        status: 'running',
        message: 'the node is up and running 🚀!',
      });
    } else {
      nodeInstance.kill('SIGINT');
      nodeInstance = null;
      mainWindow.webContents.send('node-status', {
        id: -1,
        timestamp: new Date().getTime(),
        status: 'idle',
        message: 'the node has been successfully stopped ✅!',
      });
      return;
    }

    const nodeBinary = nodeBinaries[process.platform];
    const binaryPath = path.join(
      directory,
      nodeBinary.replace('.tar.gz', '').replace('.zip', '')
    );

    const getCurrentTip = () => {
      if (!nodeRunning) {
        mainWindow.webContents.send('node-status', {
          id: -1,
          timestamp: new Date().getTime(),
          status: 'idle',
          message: 'the node has been successfully stopped ✅!',
        });
        nodeInstance.kill('SIGINT');
        nodeInstance = null;
        clearInterval(intervalId);
        intervalId = -1;
        return;
      }
      const databaseDirectory = path.join(directory, `${network}-db`);
      const args = ['query', 'tip'];

      if (network === 'mainnet') {
        args.push('--mainnet');
      } else {
        args.push('--testnet-magic');
        args.push(networks[network].protocolMagic);
      }

      const child = spawn(path.join(binaryPath, 'cardano-cli'), args, {
        env: {
          CARDANO_NODE_SOCKET_PATH: path.join(databaseDirectory, 'node.socket'),
        },
      });

      child.stdout.on('data', (data) => {
        try {
          const response = JSON.parse(data);
          mainWindow.webContents.send('node-status', {
            id: 2,
            timestamp: new Date().getTime(),
            status: 'running',
            message: `node sync progress: ${response.syncProgress}%`,
          });
        } catch (error) {
          if (!(error instanceof SyntaxError)) {
            console.error(error);
          }
        }
      });

      child.stderr.on('data', (data) => {
        console.error(`stderr: ${data}`);
      });
    };

    intervalId = setInterval(() => getCurrentTip(), 5000);
  });

  ipcMain.handle('getDefaultPath', () =>
    path.join(app.getPath('home'), '.cardano-node-ui')
  );
});

app.on('window-all-closed', function () {
  if (process.platform !== 'darwin') app.quit();
});
